include/zephyr/bluetooth/classic/hid_device.h

```diff
@@ -0,0 +1,193 @@
+/** @file
+ *  @brief HID Device Protocol handling.
+ */
+
+/*
+ * Copyright 2024 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __BT_HID_DEVICE_H
+#define __BT_HID_DEVICE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/bluetooth/classic/sdp.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/l2cap.h>
+
+#define BT_HID_MAX_MTU      64
+#define HID_REPORT_DATA_LEN 64
+
+/** @brief HID event type */
+enum {
+	BT_HID_EVENT_GET_REPORT,
+	BT_HID_EVENT_SET_REPORT,
+	BT_HID_EVENT_GET_PROTOCOL,
+	BT_HID_EVENT_SET_PROTOCOL,
+	BT_HID_EVENT_INTR_DATA,
+	BT_HID_EVENT_UNPLUG,
+	BT_HID_EVENT_SUSPEND,
+	BT_HID_EVENT_EXIT_SUSPEND,
+};
+
+/** @brief HID report types in get, set, data */
+enum {
+	BT_HID_REP_TYPE_OTHER = 0,
+	BT_HID_REP_TYPE_INPUT,
+	BT_HID_REP_TYPE_OUTPUT,
+	BT_HID_REP_TYPE_FEATURE,
+};
+
+/** @brief HID handshake response code */
+enum {
+	BT_HID_HANDSHAKE_RSP_SUCCESS = 0,
+	BT_HID_HANDSHAKE_RSP_NOT_READY,
+	BT_HID_HANDSHAKE_RSP_ERR_INVALID_REP_ID,
+	BT_HID_HANDSHAKE_RSP_ERR_UNSUPPORTED_REQ,
+	BT_HID_HANDSHAKE_RSP_ERR_INVALID_PARAM,
+	BT_HID_HANDSHAKE_RSP_ERR_UNKNOWN = 14,
+	BT_HID_HANDSHAKE_RSP_ERR_FATAL,
+};
+
+/** @brief HID protocol parameters for set protocal */
+enum {
+	BT_HID_PROTOCOL_BOOT_MODE = 0,
+	BT_HID_PROTOCOL_REPORT_MODE,
+};
+
+/** @brief HID role */
+typedef enum bt_hid_role {
+	BT_HID_ROLE_ACCEPTOR,
+	BT_HID_ROLE_INITIATOR
+} __packed bt_hid_role_t;
+
+/** @brief HID seesion role */
+typedef enum bt_hid_session_role {
+	BT_HID_SESSION_ROLE_CTRL,
+	BT_HID_SESSION_ROLE_INTR
+} __packed bt_hid_session_role_t;
+
+/** @brief HID session structure. */
+struct bt_hid_session {
+	struct bt_l2cap_br_chan br_chan;
+	bt_hid_session_role_t role; /* As ctrl or intr */
+};
+
+/** @brief HID Device structure. */
+struct bt_hid {
+	struct bt_conn *conn;
+	struct bt_hid_session ctrl_session;
+	struct bt_hid_session intr_session;
+	bt_hid_role_t init_role; /* As initial or accept */
+	uint8_t hid_state;
+	uint8_t hid_unplug;
+	struct net_buf *buf;
+};
+
+/** @brief HID Device report structure. */
+struct bt_hid_report {
+	struct bt_conn *conn;
+	uint8_t report_type;
+	uint8_t has_size: 1;
+	uint8_t reserved: 3;
```

Comment:
> This makes no sense to me. You have a pointer (`conn`) inside this struct, so it can't be a protocol struct. However then you also have a `reserved` bitfield there. What's going on? :)

Done